HW1/P4/P4.py

```diff
@@ -0,0 +1,57 @@
+import re
+from P4_bfs import bfs
+
+split_quote = re.compile(r'","')
+
+data = sc.textFile('source.csv')
+
+def clean_text(text): 
+	stripped_first_last = text[1:len(text) - 1]
+
+	# make sure that our strings are at least 2 chars 
+	assert stripped_first_last != ''
+
+	return tuple(split_quote.split(stripped_first_last)) 
+
+char_com = data.map(clean_text)
+com_char = char_com.map(lambda (x,y): (y,x)) 
+com_all_chars = com_char.groupByKey()
+
+char_chars = com_char.join(com_all_chars).values().map(lambda (k,v): (k, tuple(v)))
+edge_lookup = char_chars.reduceByKey(lambda x,y : set(x) | set(y)).cache()
+
+root = 'CARTER, PEGGY'
+
+finished = sc.parallelize([(root, (0, [[root]]))])
```

Comment:
Correct me if I am wrong: You are working with two RDDs where one contains the "finished" nodes, and another with the "unfinished" nodes. You then transfer the nodes from unfinished to finished.
It might be easier to "build" the counter into the edge_lookup RDD. This saves you the trouble of working with 2 RDDs. You set the counter for each character to infinity (or some huge number), then in each iteration, you can update the counter for nodes that have been reached.